src/Compilers/CSharp/Portable/Syntax/SyntaxNormalizer.cs

```diff
@@ -391,6 +398,130 @@ private static int LineBreaksAfterSemicolon(SyntaxToken currentToken, SyntaxToke
             }
         }
 
+        private static bool NeedsSeparatorForPropertyPattern(SyntaxToken token, SyntaxToken next)
+        {
+            PropertyPatternClauseSyntax? propPattern;
+            if (token.Parent.IsKind(SyntaxKind.PropertyPatternClause))
+            {
+                propPattern = (PropertyPatternClauseSyntax)token.Parent;
+            }
+            else if (next.Parent.IsKind(SyntaxKind.PropertyPatternClause))
+            {
+                propPattern = (PropertyPatternClauseSyntax)next.Parent;
+            }
+            else
+            {
+                return false;
+            }
+
+            var tokenIsOpenBrace = token.IsKind(SyntaxKind.OpenBraceToken);
+            var nextIsOpenBrace = next.IsKind(SyntaxKind.OpenBraceToken);
+            var tokenIsCloseBrace = token.IsKind(SyntaxKind.CloseBraceToken);
+            var nextIsCloseBrace = next.IsKind(SyntaxKind.CloseBraceToken);
+
+            //inner
+            if (tokenIsOpenBrace)
+            {
+                return true;
+            }
+            if (nextIsCloseBrace)
+            {
+                return true;
+            }
+
+            if (propPattern.Parent is RecursivePatternSyntax rps)
+            {
+                //outer
+                if (nextIsOpenBrace)
+                {
+                    if (rps.Type != null || rps.PositionalPatternClause != null)
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        return false;
+                    }
+                }
+                if (tokenIsCloseBrace)
+                {
+                    if (rps.Designation is null)
+                    {
+                        return false;
+                    }
+                    else
+                    {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        private static bool NeedsSeparatorForPositionalPattern(SyntaxToken token, SyntaxToken next)
+        {
+            PositionalPatternClauseSyntax? posPattern;
+            if (token.Parent.IsKind(SyntaxKind.PositionalPatternClause))
```

Comment:
Same comment about `IsKind` followed by a cast on two nodes vs `as` not being a perf concern.

Reply:
`PositionalPatternClauseSyntax` is a `sealed` class - so I think the `is` check should just do a null check and a comparison of the VMT pointer. With IsKind we compare simply the stored integer and omit the null check. The performance difference should be not that big.
For non-sealed classes it pays maybe a perfomance penalty.
But for those we can't do a simple IsKind check (e.g. Statement Syntax).

Reply:
I have seen that there are still some issues. E.g. `Foo() swtich { ...}` is normalized to `Foo()swtich { ...}`.
Or property-initializers are not normalized well - all in a single line - which is fine for collection initializers.

But those changes I would put in another PR.